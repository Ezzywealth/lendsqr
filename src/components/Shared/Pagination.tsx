import React from 'react';
import { MdKeyboardArrowRight, MdKeyboardArrowLeft } from 'react-icons/md';
import { PaginationProps } from '../../interfaces/typings';
import PaginationBtns from './PaginationBtns';

const Pagination = ({ currentPage, totalPages, itemsPerPage, totalItems, onPageChange }: PaginationProps) => {
	const { renderPageButtons, handlePageChange } = PaginationBtns({ currentPage, totalPages, itemsPerPage, totalItems, onPageChange });

	return (
		<div className='pagination-container'>
			<div className='pagination-info'>
				<p>Showing</p>
				<select
					value={currentPage}
					onChange={(e) => {
						onPageChange(+e.target.value);
					}}
					className='select-field'>
					{Array.from({ length: totalPages }, (_, index) => (
						<option key={index + 1} value={index + 1} data-testid={`option-${index + 1}`}>
							{index + 1}
						</option>
					))}
				</select>
				out of {totalPages}
			</div>
			{totalPages > 1 && (
				<div className='pagination-pages'>
					<span className='pagination-arrow'>
						<MdKeyboardArrowLeft size={25} onClick={() => handlePageChange(currentPage - 1)} />
					</span>
					{renderPageButtons()}
					<span className='pagination-arrow'>
						<MdKeyboardArrowRight size={25} onClick={() => handlePageChange(currentPage + 1)} />
					</span>
				</div>
			)}
		</div>
	);
};

export default Pagination;
