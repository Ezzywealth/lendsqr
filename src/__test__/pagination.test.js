/* eslint-disable testing-library/no-unnecessary-act */
/* eslint-disable testing-library/no-node-access */
/* eslint-disable jest/no-conditional-expect */
import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Pagination from '../components/Shared/Pagination';

const defaultProps = {
	currentPage: 1,
	totalPages: 10,
	itemsPerPage: 10,
	totalItems: 100,
	onPageChange: jest.fn(),
};

// Mock the Redux store and related dependencies
jest.mock('../Redux/store.ts', () => ({
	...jest.requireActual('../Redux/store.ts'),
	useAppDispatch: jest.fn(),
}));

// Mock the userSlice module
jest.mock('../Redux/slices/userSlice', () => ({
	...jest.requireActual('../Redux/slices/userSlice'),
	initialState: {
		currentPage: 1,
		itemsPerPage: 10,
		totalItems: 100,
		totalPages: 10,
	},
	setItemsPerPage: jest.fn(),
}));

describe('Pagination component', () => {
	test('renders Pagination component', () => {
		const { container } = render(<Pagination {...defaultProps} />);
		expect(container).toMatchSnapshot();
	});
	it('renders Pagination correctly', () => {
		render(<Pagination {...defaultProps} />);
		// Assert
		// Find the element with the "Showing" text within the pagination container
		expect(screen.getByText('Showing', { selector: 'p' })).toBeInTheDocument();
	});

	it('handles page change on button click', () => {
		render(<Pagination {...defaultProps} />);
		// Act
		const buttonWithText2 = screen.getByText('2', { selector: 'button' });
		fireEvent.click(buttonWithText2);
		// Assert
		expect(defaultProps.onPageChange).toHaveBeenCalledWith(2);
	});

	// ...

	it('handles items per page change', () => {
		render(<Pagination {...defaultProps} />);

		// Act

		const option1 = screen.getByTestId('option-1');
		expect(option1).toBeInTheDocument();
	});

	it('renders correct number of page buttons and ellipsis', () => {
		const totalPages = 10; // Change this to the desired total number of pages
		// Act
		render(<Pagination currentPage={1} totalPages={totalPages} itemsPerPage={10} totalItems={100} onPageChange={() => {}} />);
		// Assert
		const pageButtons = screen.getAllByRole('button');

		const ellipsis = screen.getByText('...');

		// Assert that the correct number of page buttons is rendered
		expect(pageButtons).toHaveLength(5);

		// Assert that the ellipsis is visible when there are more than 5 pages
		if (totalPages > 5) {
			expect(ellipsis).toBeInTheDocument();
		} else {
			// Assert that the ellipsis is not present when there are 5 or fewer pages
			expect(ellipsis).not.toBeInTheDocument();
		}
	});

	// Add more test cases as needed
});
