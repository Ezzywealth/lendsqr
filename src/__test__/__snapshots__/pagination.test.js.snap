// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination component renders Pagination component 1`] = `
<div>
  <div
    class="pagination-container"
  >
    <div
      class="pagination-info"
    >
      <p>
        Showing
      </p>
      <select
        class="select-field"
      >
        <option
          data-testid="option-1"
          value="1"
        >
          1
        </option>
        <option
          data-testid="option-2"
          value="2"
        >
          2
        </option>
        <option
          data-testid="option-3"
          value="3"
        >
          3
        </option>
        <option
          data-testid="option-4"
          value="4"
        >
          4
        </option>
        <option
          data-testid="option-5"
          value="5"
        >
          5
        </option>
        <option
          data-testid="option-6"
          value="6"
        >
          6
        </option>
        <option
          data-testid="option-7"
          value="7"
        >
          7
        </option>
        <option
          data-testid="option-8"
          value="8"
        >
          8
        </option>
        <option
          data-testid="option-9"
          value="9"
        >
          9
        </option>
        <option
          data-testid="option-10"
          value="10"
        >
          10
        </option>
      </select>
      out of 
      10
    </div>
    <div
      class="pagination-pages"
    >
      <span
        class="pagination-arrow"
      >
        <svg
          fill="currentColor"
          height="25"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="25"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M15.41 16.59L10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
          />
        </svg>
      </span>
      <button
        class="pagination-page active"
      >
        1
      </button>
      <button
        class="pagination-page "
      >
        2
      </button>
      <button
        class="pagination-page "
      >
        3
      </button>
      <span
        class="pagination-dots"
      >
        ...
      </span>
      <button
        class="pagination-page "
      >
        9
      </button>
      <button
        class="pagination-page "
      >
        10
      </button>
      <span
        class="pagination-arrow"
      >
        <svg
          fill="currentColor"
          height="25"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="25"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
          />
        </svg>
      </span>
    </div>
  </div>
</div>
`;
